OutOfSchool/OutOfSchool.WebApi/Services/CurrentUserService.cs

```diff
@@ -0,0 +1,134 @@
+        this.parentService = parentService;
+
+        // TODO: add caching as this service is going to be called everytime use makes requests
+        // TODO: so probably can cache extended user info for quick validations
+    }
+
+    public string UserId => user?.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Sub) ?? string.Empty;
+
+    public bool IsInRole(Role role) => role switch
+    {
+        Role.Provider => user?.IsInRole("provider") ?? false,
+        Role.Parent => user?.IsInRole("parent") ?? false,
+        Role.TechAdmin => user?.IsInRole("techadmin") ?? false,
+        Role.MinistryAdmin => user?.IsInRole("ministryadmin") ?? false,
+        _ => false
+    };
+
+    // TODO: in code we check multiple permissions (for example parent and provider can edit application status)
+    // TODO: either separate method calls with IsInRole check, or array like here?
+    public async Task UserHasRights(params IUserType[] userTypes)
+    {
+        // TODO: double check role?
+        var userHasRights = false;
+        if (user?.Identity?.IsAuthenticated ?? false)
+        {
+            var parent = userTypes.Select(t => t as ParentType).FirstOrDefault(p => p is not null);
+            var provider = userTypes.Select(t => t as ProviderType).FirstOrDefault(p => p is not null);
+            var providerAdmin = userTypes.Select(t => t as ProviderAdminType).FirstOrDefault(p => p is not null);
+
+            var result = await Task.WhenAll(
+                new List<Task<bool>>
+                    {
+                        UserHasRights(parent),
+                        UserHasRights(provider),
+                        UserHasRights(providerAdmin),
+                    }
+                    .Select(Execute));
+            userHasRights = result.Aggregate(false, (acc, next) => acc || next);
+        }
+
+        if (!userHasRights)
+        {
+            // TODO: copied from existing. Need custom exception
+            throw new ArgumentException("User has no rights to perform operation");
+        }
+    }
+
+    private static async Task<bool> Execute(Task<bool> original)
+    {
+        try
+        {
+            return await original;
+        }
+        catch (Exception)
+        {
+            // TODO: Logging?
+            return false;
+        }
+    }
+
+    private Task<bool> UserHasRights<T>(T? userType)
+        where T : IUserType
+        => userType switch
+        {
+            ParentType parent => ParentHasRights(parent.parentId),
+            ProviderType provider => ProviderHasRights(provider.providerAdminId),
+            ProviderAdminType providerAdmin => ProviderAdminHasRights(
+                providerAdmin.providerId,
+                providerAdmin.workshopId),
+            _ => Task.FromResult(false)
+        };
+
+    private async Task<bool> ParentHasRights(Guid parentId)
+    {
+        var parent = await parentService.GetByUserId(UserId).ConfigureAwait(false);
```

Comment:
Perf consideration: Here we take and map the whole object just to check if id is same. Isn't it better to move this check on repo/SQL level? 

I know that it slightly breaks used abstractions, but as check is pretty simple and we could gain perf boost, probably, why not do it?

Reply:
I was thinking about having some of user information cached in Redis, because:
1. These checks gonna happen with every user request to protected operations. 
2. Most Id's (e.g., userid, parent or provider id) will never change, so safe to cache.

Was thinking about `~5 min` cache time for this info as according to some statistic i found in google - average session duration on a website in around 3 minutes.